File: src/morphic/resizer/scanner.py
"""
Background scan job management for the resizer module.

Discovers images in a folder and resizes them in a background thread.
"""

from __future__ import annotations

import logging
import os
import threading
import time
import uuid
from dataclasses import dataclass, field

from morphic.shared.constants import IMAGE_EXTENSIONS
from morphic.shared.utils import (
    find_files_by_extension,
    format_duration,
    format_file_size,
)

logger = logging.getLogger(__name__)


@dataclass
class ScanJob:
    """Represents a running or completed resize job."""

    id: str
    folder: str
    width: int
    height: int
    mode: str
    output_folder: str | None = None
    bg_color: str = "#000000"
    quality: int = 90
    status: str = "pending"
    progress: float = 0.0
    message: str = ""
    error: str | None = None
    total_files: int = 0
    processed_files: int = 0
    errors: list[dict] = field(default_factory=list)
    results: list[dict] = field(default_factory=list)
    started_at: float = 0.0
    finished_at: float = 0.0


# ── Job Registry ───────────────────────────────────────────────────────────

_jobs: dict[str, ScanJob] = {}
_lock = threading.Lock()


def get_job(job_id: str) -> ScanJob | None:
    """Retrieve a resize job by ID."""
    with _lock:
        return _jobs.get(job_id)


def start_job(
    folder: str,
    width: int,
    height: int,
    mode: str,
    output_folder: str | None = None,
    bg_color: str = "#000000",
    quality: int = 90,
) -> str:
    """Create and launch a new resize job. Returns the job ID."""
    job_id = str(uuid.uuid4())[:8]
    job = ScanJob(
        id=job_id,
        folder=folder,
        width=width,
        height=height,
        mode=mode,
        output_folder=output_folder,
        bg_color=bg_color,
        quality=quality,
    )
    with _lock:
        _jobs[job_id] = job

    thread = threading.Thread(target=_run_resize, args=(job,), daemon=True)
    thread.start()
    return job_id


def _run_resize(job: ScanJob) -> None:
    """Execute the resize operation in a background thread."""
    from morphic.resizer.operations import resize_image

    try:
        job.status = "scanning"
        job.started_at = time.time()
        job.message = f"Scanning folder: {job.folder}"

        paths = find_files_by_extension(job.folder, IMAGE_EXTENSIONS)
        job.total_files = len(paths)

        if not paths:
            job.status = "done"
            job.progress = 1.0
            job.finished_at = time.time()
            job.message = "No image files found."
            return

        job.status = "processing"
        for i, path in enumerate(paths):
            try:
                original_size = os.path.getsize(path)
                dest = resize_image(
                    path,
                    job.width,
                    job.height,
                    mode=job.mode,
                    output_folder=job.output_folder,
                    bg_color=job.bg_color,
                    quality=job.quality,
                )
                new_size = os.path.getsize(dest) if os.path.isfile(dest) else 0
                job.results.append(
                    {
                        "source": path,
                        "destination": dest,
                        "status": "ok",
                        "original_size": original_size,
                        "new_size": new_size,
                        "original_size_fmt": format_file_size(original_size),
                        "new_size_fmt": format_file_size(new_size),
                    }
                )
            except Exception as e:
                job.errors.append({"path": path, "error": str(e)})
                job.results.append(
                    {
                        "source": path,
                        "destination": None,
                        "status": "error",
                        "error": str(e),
                    }
                )

            job.processed_files = i + 1
            job.progress = (i + 1) / job.total_files
            job.message = (
                f"Resizing: {i + 1}/{job.total_files} "
                f"({len(job.errors)} errors)"
            )

        job.status = "done"
        job.progress = 1.0
        job.finished_at = time.time()
        elapsed = job.finished_at - job.started_at
        job.message = (
            f"Done! Resized {job.processed_files} images "
            f"in {format_duration(elapsed)}. "
            f"{len(job.errors)} error(s)."
        )

    except Exception as e:
        logger.exception("Resize job failed")
        job.status = "error"
        job.error = str(e)
        job.message = f"Error: {e}"
        job.finished_at = time.time()

File: internal/resizer/scanner.go
package resizer

import (
	"fmt"
	"os"
	"sync"
	"time"

	"github.com/exterex/morphic/internal/shared"
)

// ScanJob represents a running or completed resize job.
type ScanJob struct {
	shared.Job
	mu sync.Mutex

	Folder       string                   `json:"folder"`
	Width        int                      `json:"width"`
	Height       int                      `json:"height"`
	Mode         string                   `json:"mode"`
	OutputFolder string                   `json:"output_folder,omitempty"`
	BgColor      string                   `json:"bg_color"`
	Quality      int                      `json:"quality"`
	Total        int                      `json:"total"`
	Processed    int                      `json:"processed"`
	Results      []map[string]interface{} `json:"results,omitempty"`
	Errors       []map[string]interface{} `json:"errors,omitempty"`
}

var store = shared.NewJobStore[ScanJob]()

func init() {
	store.StartCleanup(30*time.Minute, func(j *ScanJob) time.Time {
		return j.DoneAt
	})
}

// StartJob creates and launches a new resize job.
func StartJob(folder string, width, height int, mode, outputFolder, bgColor string, quality int) string {
	job := &ScanJob{
		Job:          shared.NewJob(),
		Folder:       folder,
		Width:        width,
		Height:       height,
		Mode:         mode,
		OutputFolder: outputFolder,
		BgColor:      bgColor,
		Quality:      quality,
	}
	job.Status = shared.JobStatusRunning
	store.Set(job.ID, job)
	go runResize(job)
	return job.ID
}

// GetJob retrieves a job by ID.
func GetJob(id string) (*ScanJob, bool) {
	return store.Get(id)
}

func runResize(job *ScanJob) {
	defer func() {
		if r := recover(); r != nil {
			job.mu.Lock()
			job.Status = shared.JobStatusFailed
			job.Error = fmt.Sprintf("%v", r)
			job.DoneAt = time.Now()
			job.mu.Unlock()
		}
	}()

	job.mu.Lock()
	job.Message = fmt.Sprintf("Scanning folder: %s", job.Folder)
	job.mu.Unlock()

	files, err := shared.FindImageFiles(job.Folder)
	if err != nil {
		job.mu.Lock()
		job.Status = shared.JobStatusFailed
		job.Error = err.Error()
		job.DoneAt = time.Now()
		job.mu.Unlock()
		return
	}

	job.mu.Lock()
	job.Total = len(files)
	job.mu.Unlock()

	if len(files) == 0 {
		job.mu.Lock()
		job.Status = shared.JobStatusDone
		job.Progress = 1.0
		job.DoneAt = time.Now()
		job.Message = "No image files found."
		job.mu.Unlock()
		return
	}

	errCount := 0
	for i, f := range files {
		origSize := f.Size
		dest, resizeErr := ResizeImage(
			f.Path, job.Width, job.Height, job.Mode,
			job.OutputFolder, job.BgColor, job.Quality, "",
		)

		result := map[string]interface{}{
			"source": f.Path,
		}

		if resizeErr != nil {
			errCount++
			result["destination"] = nil
			result["status"] = "error"
			result["error"] = resizeErr.Error()

			job.mu.Lock()
			job.Errors = append(job.Errors, map[string]interface{}{
				"path":  f.Path,
				"error": resizeErr.Error(),
			})
			job.mu.Unlock()
		} else {
			var newSize int64
			if info, err := os.Stat(dest); err == nil {
				newSize = info.Size()
			}
			result["destination"] = dest
			result["status"] = "ok"
			result["original_size"] = origSize
			result["new_size"] = newSize
			result["original_size_fmt"] = shared.FormatFileSize(origSize)
			result["new_size_fmt"] = shared.FormatFileSize(newSize)
		}

		job.mu.Lock()
		job.Results = append(job.Results, result)
		job.Processed = i + 1
		job.Progress = float64(i+1) / float64(job.Total)
		job.Message = fmt.Sprintf("Resizing: %d/%d (%d errors)", i+1, job.Total, errCount)
		job.mu.Unlock()
	}

	job.mu.Lock()
	job.Status = shared.JobStatusDone
	job.Progress = 1.0
	job.DoneAt = time.Now()
	elapsed := job.DoneAt.Sub(job.StartedAt).Seconds()
	job.Message = fmt.Sprintf("Done! Resized %d images in %s. %d error(s).",
		job.Processed, shared.FormatDuration(elapsed), errCount)
	job.mu.Unlock()
}
